maven-core/src/main/java/org/apache/maven/internal/transformation/ConsumerPomArtifactTransformer.java

```diff
@@ -77,13 +80,34 @@ public void injectTransformedArtifacts(MavenProject project, RepositorySystemSes
                 Files.createDirectories(buildDir);
                 generatedFile = Files.createTempFile(buildDir, CONSUMER_POM_CLASSIFIER, "pom");
             }
+
+            removeOldConsumerPomFiles(generatedFile);
```

Comment:
Can we move the call to `removeOldConsumerPomFiles` before generating the new file and remove them all instead of having to pass the newly created file as an argument ?

Reply:
Also, I think this removing *all* matching files from the temporary directory is problematic.  If we have two concurrent builds running, one of the files will be lost.  This may be done only when `buildDirectory != null`.

Reply:
> I think it's easier to remove all files, then generate the new one...

Fair point. It makes it simpler and clearer, nice catch!

> Also, I think this removing _all_ matching files from the temporary directory is problematic. If we have two concurrent builds running, one of the files will be lost. This may be done only when `buildDirectory != null`.

It would be a problem. I don't think we can safely delete any files in that case without introducing too much complexity, so the behaviour would be to just let the files live in the temp dir and _maybe_ get cleaned up by the OS. Some do and some don't.

I didn't know the temp dir would/could be used as build directory (I just looked and the files were indeed there). Out of curiosity; when would this be the case?

Reply:
> > Also, I think this removing _all_ matching files from the temporary directory is problematic. If we have two concurrent builds running, one of the files will be lost. This may be done only when `buildDirectory != null`.
> 
> It would be a problem. I don't think we can safely delete any files in that case without introducing too much complexity, so the behaviour would be to just let the files live in the temp dir and _maybe_ get cleaned up by the OS. Some do and some don't.

There may be a better alternative.  I wonder if we could just add a `@PreDestroy` method that would keep a list of generated files (be it in the temp directory or in the build directory) and delete them when the bean is destroy...

> I didn't know the temp dir would/could be used as build directory (I just looked and the files were indeed there). Out of curiosity; when would this be the case?

Actually, I did not know either, I just noticed it while reviewing this PR.

Reply:
@Giovds have a look at https://github.com/gnodet/maven/commit/11009e9f6921657e31384bd9838d5ab1edb1aacc, could you double check that it works for you ?

Reply:
> @Giovds have a look at https://github.com/gnodet/maven/commit/11009e9f6921657e31384bd9838d5ab1edb1aacc, could you double check that it works for you ?

Cool, TIL! Just verified it and it does work for me. The files are cleaned up. 

Do note that the final result means there will be no consumer*pom file in the build dir whatsoever. I don't know if it was kept for a reason. Perhaps the file is/will be (re)used for build-consumer features beyond a single run? If it's not and it's a temp file, why not always use the temp dir e.g.?

\* Without to much research: * Seems like (in this class at least) the `CONSUMER_POM_CLASSIFIER` is private so not used by other classes and the build dir is never queried for the file, only in-memory artifact lists?

Reply:
> > @Giovds have a look at [gnodet@11009e9](https://github.com/gnodet/maven/commit/11009e9f6921657e31384bd9838d5ab1edb1aacc), could you double check that it works for you ?
> 
> Cool, TIL! Just verified it and it does work for me. The files are cleaned up.
> 
> Do note that the final result means there will be no consumer*pom file in the build dir whatsoever. I don't know if it was kept for a reason. Perhaps the file is/will be (re)used for build-consumer features beyond a single run? If it's not and it's a temp file, why not always use the temp dir e.g.?
> 
> * Without to much research: * Seems like (in this class at least) the `CONSUMER_POM_CLASSIFIER` is private so not used by other classes and the build dir is never queried for the file, only in-memory artifact lists?

Yes, I think it was written to disk because it was needed as a file and there was no easy way to delete it at the end of the process, but they're not supposed to outlive the session.